gooddata-pipelines/tests/provisioning/entities/users/test_permissions.py

```diff
@@ -387,3 +418,132 @@ def mock_upstream_perms(ws_id: str) -> CatalogDeclarativeWorkspacePermissions:
     if ws_id not in UPSTREAM_WS_PERMISSIONS:
         raise NotFoundException(404)
     return UPSTREAM_WS_PERMISSIONS[ws_id]
+
+
+@pytest.fixture
+def permission_provisioner(mocker: MockerFixture) -> PermissionProvisioner:
+    provisioner_instance = PermissionProvisioner.create(
+        host="https://localhost:3000", token="token"
+    )
+
+    # Patch the API
+    mocker.patch.object(provisioner_instance, "_api", return_value=None)
+
+    return provisioner_instance
+
+
+def parse_expected_permissions(
+    raw_data: dict,
+) -> dict[str, list[CatalogDeclarativeSingleWorkspacePermission]]:
+    expected_result: dict[
+        str, list[CatalogDeclarativeSingleWorkspacePermission]
+    ] = {}
+    for workspace_id, workspace_permissions in raw_data.items():
+        expected_permissions = []
+        for permission in workspace_permissions:
+            expected_permissions.append(
+                CatalogDeclarativeSingleWorkspacePermission(
+                    name=permission["name"],
+                    assignee=CatalogAssigneeIdentifier(
+                        id=permission["assignee_id"],
+                        type=permission["assignee_type"],
+                    ),
+                )
+            )
+        expected_result[workspace_id] = expected_permissions
+    return expected_result
+
+
+@pytest.mark.parametrize(
+    ("source_data_path", "expected_data_path", "load_method"),
+    [
+        (
+            "permissions_input_full_load.json",
+            "permissions_expected_full_load.json",
+            "full_load",
+        ),
+        (
+            "permissions_input_incremental_load.json",
+            "permissions_expected_incremental_load.json",
+            "incremental_load",
+        ),
+    ],
+)
+def test_permission_provisioner(
+    source_data_path: str,
+    expected_data_path: str,
+    load_method: Literal["incremental_load", "full_load"],
+    permission_provisioner: PermissionProvisioner,
+    mocker: MockerFixture,
+) -> None:
+    source_data: list[dict] = []
+    incremental_load_data: list[PermissionIncrementalLoad] = []
+    full_load_data: list[PermissionFullLoad] = []
+
+    # Load existing upstream permissions
+    EXISTING_UPSTREAM_PERMISSIONS_PATH = (
+        f"{TEST_DATA_SUBDIR}/existing_upstream_permissions.json"
+    )
+    with open(EXISTING_UPSTREAM_PERMISSIONS_PATH, "r") as f:
+        raw_existing_upstream_permissions = json.load(f)
```

Comment:
No need to change it now, but consider using `orjson` in the future. It is a fast alternative to JSON written in Rust.